sql-serializer/src/main/java/io/cdap/pipeline/sql/serializer/bigquery/BigQuerySQLSerializer.java

```diff
@@ -0,0 +1,247 @@
+      throw new IllegalArgumentException("Temporary tables cannot be fully qualified");
+    }
+    StringBuilder b = new StringBuilder();
+    b.append("CREATE TEMPORARY TABLE ");
+    appendTableSQL(b, temporaryTable, false);
+    b.append(" AS ");
+    appendQuerySQL(b, query, false);
+    return b.toString();
+  }
+
+  private StringBuilder appendQuerySQL(StringBuilder b, StructuredQuery query, boolean isSubquery) {
+    // Keep track of which columns were defined in the current nesting level since they may also be used in filter
+    HashSet<Column> definedColumns = new HashSet<>();
+    if (isSubquery) {
+      b.append('(');
+    } else if (query.hasAlias()) {
+      throw new IllegalArgumentException("Cannot have an alias on the main outer query");
+    }
+    b.append("SELECT ");
+    for (int i = 0; i < query.getColumns().size(); i++) {
+      Column col = query.getColumns().get(i);
+      if (col.hasAlias()) {
+        definedColumns.add(col);
+      }
+      appendColumnSQL(b, col, true, false);
+      if (i < query.getColumns().size() - 1) {
+        b.append(", ");
+      }
+    }
+    b.append(" FROM ");
+    appendQueryableSQL(b, query.getQueryFrom());
+    if (query.getFilter() != null) {
+      b.append(" WHERE ");
+      appendFilterSQL(b, query.getFilter(), definedColumns);
+    }
+    if (isSubquery) {
+      b.append(')');
+      if (query.hasAlias()) {
+        b.append(" AS ");
+        b.append(query.getAlias());
+      }
+    } else {
+      b.append(';');
+    }
+    return b;
+  }
+
+  private StringBuilder appendColumnSQL(StringBuilder b, Column c,
+                                        boolean defineAliasIfExists, boolean useAliasIfExists) {
+    if (defineAliasIfExists && useAliasIfExists) {
+      throw new IllegalArgumentException("Unable to use and define an alias at the same time");
+    }
+    if (c.hasSource() && c.getSource().hasAlias()) {
+      b.append(c.getSource().getAlias());
+      b.append('.');
+    }
+    if (c.hasAlias() && useAliasIfExists) {
+      b.append(c.getAlias());
+    } else {
+      b.append(c.getName());
+    }
+    if (c.hasAlias() && defineAliasIfExists) {
+      b.append(" AS ");
+      b.append(c.getAlias());
+    }
+    return b;
+  }
+
+  private StringBuilder appendTableSQL(StringBuilder b, Table t, boolean defineAliasIfExists) {
+    if (!t.hasAlias() || defineAliasIfExists) {
+      b.append('`');
+      if (t.hasProject()) {
+        b.append(t.getProjectName());
+        b.append('.');
+      }
+      if (t.hasDatabase()) {
+        b.append(t.getDatabaseName());
+        b.append('.');
+      }
+      b.append(t.getTableName());
+      b.append('`');
+      if (t.hasAlias() && defineAliasIfExists) {
+        b.append(" AS ");
+        b.append(t.getAlias());
+      }
+    } else {
+      b.append('`');
+      b.append(t.getAlias());
+      b.append('`');
+    }
+    return b;
+  }
+
+  private StringBuilder appendQueryableSQL(StringBuilder b, Queryable s) {
```

Comment:
This method is where I think it would make more sense to wrap the result in parentheses and potentially alias it, as you know it is always within a query already.

Reply:
Fixed